js/runner.go

```diff
@@ -274,6 +274,13 @@ func forceHTTP1() bool {
 
 // Setup runs the setup function if there is one and sets the setupData to the returned value
 func (r *Runner) Setup(ctx context.Context, out chan<- metrics.SampleContainer) error {
+	if !r.IsExecutable(consts.SetupFn) {
+		// do not init a new transient VU or execute setup() if it wasn't
+		// actually defined and exported in the script
+		r.preInitState.Logger.Debugf("%s() is not defined or not exported, skipping...", consts.SetupFn)
```

Comment:
Like I mentioned on Slack, I'd rather move this check to where `Runner.Setup()` is called. There's no point in calling a function that will no-op, if the same check could be done by the caller. Besides being straightforward to read, we wouldn't end up with this confusing log:

```
DEBU[0000] Running setup()                               phase=local-execution-scheduler-run
DEBU[0000] setup() is not defined or not exported, skipping...
```

If `setup()` is not exported or defined, what are we running...?

Additionally, does it make sense to [change the execution status and progressbar](https://github.com/grafana/k6/blob/56fcad04e33db6359322f57534be474c861b3932/core/local/local.go#L424-L425) if we don't end up running it?

Also, wouldn't returning `nil` here be a minor breaking change for the HTTP API? [We run `Setup()` on `POST /v1/setup`](https://github.com/grafana/k6/blob/56fcad04e33db6359322f57534be474c861b3932/api/v1/setup_teardown_routes.go#L78), and previously we would return a 500 error if `setup` is not found, whereas now it returns a 200 without actually running it. It's questionable whether 500 is a good fit (a 4xx code might be more suitable), but a 2xx definitely wouldn't make sense if we didn't run anything. I don't think this would actually break anything, but it's one more reason to not do this here.

So I would definitely prefer that this check was added [here](https://github.com/grafana/k6/blob/56fcad04e33db6359322f57534be474c861b3932/core/local/local.go#L422) instead. If we want to keep this log, then it should be done from there as well. Though I wonder if we need it, since we also don't log if `setup()` is not run because of `--no-setup`.

Reply:
> Also, wouldn't returning `nil` here be a minor breaking change for the HTTP API? [We run `Setup()` on `POST /v1/setup`](https://github.com/grafana/k6/blob/56fcad04e33db6359322f57534be474c861b3932/api/v1/setup_teardown_routes.go#L78), and previously we would return a 500 error if `setup` is not found, whereas now it returns a 200 without actually running it.

Did you check that? Because it shouldn't have returned 500 if there was no `setup()` before, `runPart()` should have returned `goja.Undefined`: https://github.com/grafana/k6/blob/70fa2b8d2884c86976f18e43f7074f1a7a8f645c/js/runner.go#L499-L502
And `js.Runner.Setup()` should have just returned `nil` for an error: https://github.com/grafana/k6/blob/70fa2b8d2884c86976f18e43f7074f1a7a8f645c/js/runner.go#L284-L288

500 error in the `POST /v1/setup` when no there wasn't as `setup()` function would be a bug.

Reply:
To respond to the rest of the comment, I have two problems with checking for this in the `ExecutionScheduler`. The first one is that it will make an already long and complex method even longer and more complex.

The second and bigger problem is that, conceptually, it's not any of the `ExecutionScheduler`'s concern how a "setup" call is implemented. The `lib.Runner` interface, flawed as it is, doesn't expose that information, it just has this method (notice the comment): https://github.com/grafana/k6/blob/70fa2b8d2884c86976f18e43f7074f1a7a8f645c/lib/runner.go#L59-L60

The `js.Runner` implementation expects the script to has an exported method called `setup()` that behaves in a specific way. The `testutils.MiniRunner` implementation of the same interface, however, just has this: https://github.com/grafana/k6/blob/70fa2b8d2884c86976f18e43f7074f1a7a8f645c/lib/testutils/minirunner/minirunner.go#L23

Another future implementation (e.g. lua, https://github.com/grafana/k6/issues/751, https://github.com/grafana/k6/issues/2562, whatever) could have a completely different way of handling "setup".

So, my suggestion would actually be to move this log to the JS runner, so there are no conflicting logs: https://github.com/grafana/k6/blob/70fa2b8d2884c86976f18e43f7074f1a7a8f645c/core/local/local.go#L423

Reply:
The [latest force-push](https://github.com/grafana/k6/compare/876e8174d8f4de119de69656f2604316e54c07f5..10c7581aca2d1255ad8efeba1670d0f3fe70068f) cleans up the conflicting logging.

Reply:
Ah, I misread the `runPart()` handling of this. :facepalm: And no, I didn't actually test it. You're right, it wouldn't change. :+1:

I don't quite see that this additional check would add much complexity to `ExecutionScheduler.Run()`, but see your point about it being conceptually specific to the JS implementation, however far away those alternative implementations might be. I suppose that would be the argument for changing the execution status and progress bar as well(?).

Yeah, the logging makes more sense in `Setup()` then.